packages/react-dom/src/server/ReactDOMServerFormatConfig.js

```diff
@@ -264,6 +283,25 @@ export function getChildFormatContext(
     return createFormatContext(HTML_MODE, null);
   }
   if (parentContext.insertionMode === ROOT_HTML_MODE) {
+    switch (type) {
+      case 'html': {
+        return parentContext;
+      }
+      case 'head':
+      case 'title':
+      case 'base':
+      case 'link':
+      case 'style':
+      case 'meta':
+      case 'script':
+      case 'noscript':
+      case 'template': {
+        break;
+      }
+      default: {
+        parentContext.preambleOpen = false;
```

Comment:
These are generally not mutable and I'm not sure this makes sense. What is this flag trying accomplish?

E.g. does this still make sense for:

```
<html>
  <AsyncHead />
   <body />
</html>
```

Where the head loads later than body. The preamble is still kind of open then isn't it?

Reply:
hmm yeah that's a problem. 

```
<html>
  <body />
  <head />
</html>
```

is also a problem though. The intent was to basically say we only use the preamble if we've not pushed anything yet but if we have (because of a poorly ordered head) we'd just encode it like regular chunks.

if we have

```
<html>
  <AsyncNoOutput />
  <AsyncHead />
  <AsyncBody />
</html>
```
It is very ambiguous whether the head use the preamble. If the first component outputs a <div> or even a <link> we're not able to have the <head> use the preamble. And while the render is probably just plain messed up but we could still emit resources after the <html> open tag. But if the first component has no output then the head one should use the preamble.

I think I may have to model the preamble differently and figure out what is included in the flush... which I was trying to avoid

Reply:
As discussed, we'll support async head but you still need to position it properly as a child of html (i.e. first child) otherwise your app will break on hydration. we'll tighten up validation on server in a follow up